shell/platform/linux/fl_value.cc

```diff
@@ -0,0 +1,550 @@
+    }
+    case FL_VALUE_TYPE_INT64_LIST: {
+      if (fl_value_get_length(a) != fl_value_get_length(b))
+        return false;
+      const int64_t* values_a = fl_value_get_int64_list(a);
+      const int64_t* values_b = fl_value_get_int64_list(b);
+      for (size_t i = 0; i < fl_value_get_length(a); i++) {
+        if (values_a[i] != values_b[i])
+          return false;
+      }
+      return true;
+    }
+    case FL_VALUE_TYPE_FLOAT_LIST: {
+      if (fl_value_get_length(a) != fl_value_get_length(b))
+        return false;
+      const double* values_a = fl_value_get_float_list(a);
+      const double* values_b = fl_value_get_float_list(b);
+      for (size_t i = 0; i < fl_value_get_length(a); i++) {
+        if (values_a[i] != values_b[i])
+          return false;
+      }
+      return true;
+    }
+    case FL_VALUE_TYPE_LIST: {
+      if (fl_value_get_length(a) != fl_value_get_length(b))
+        return false;
+      for (size_t i = 0; i < fl_value_get_length(a); i++) {
+        if (!fl_value_equal(fl_value_list_get_value(a, i),
+                            fl_value_list_get_value(b, i)))
+          return false;
+      }
+      return true;
+    }
+    case FL_VALUE_TYPE_MAP: {
+      if (fl_value_get_length(a) != fl_value_get_length(b))
+        return false;
+      for (size_t i = 0; i < fl_value_get_length(a); i++) {
+        FlValue* key = fl_value_map_get_key(a, i);
+        FlValue* value_b = fl_value_map_lookup(b, key);
+        if (value_b == nullptr)
+          return false;
+        FlValue* value_a = fl_value_map_get_value(a, i);
+        if (!fl_value_equal(value_a, value_b))
+          return false;
+      }
+      return true;
+    }
+  }
+}
+
+G_MODULE_EXPORT void fl_value_list_add(FlValue* self, FlValue* value) {
+  g_return_if_fail(self->type == FL_VALUE_TYPE_LIST);
+  g_return_if_fail(value != nullptr);
+
+  fl_value_list_add_take(self, fl_value_ref(value));
+}
+
+G_MODULE_EXPORT void fl_value_list_add_take(FlValue* self, FlValue* value) {
+  g_return_if_fail(self->type == FL_VALUE_TYPE_LIST);
+  g_return_if_fail(value != nullptr);
+
+  FlValueList* v = reinterpret_cast<FlValueList*>(self);
+  g_ptr_array_add(v->values, value);
+}
+
+G_MODULE_EXPORT void fl_value_map_set(FlValue* self,
+                                      FlValue* key,
+                                      FlValue* value) {
+  g_return_if_fail(self->type == FL_VALUE_TYPE_MAP);
+  g_return_if_fail(key != nullptr);
+  g_return_if_fail(value != nullptr);
+
+  fl_value_map_set_take(self, fl_value_ref(key), fl_value_ref(value));
+}
+
+G_MODULE_EXPORT void fl_value_map_set_take(FlValue* self,
+                                           FlValue* key,
+                                           FlValue* value) {
+  g_return_if_fail(self->type == FL_VALUE_TYPE_MAP);
+  g_return_if_fail(key != nullptr);
+  g_return_if_fail(value != nullptr);
+
+  FlValueMap* v = reinterpret_cast<FlValueMap*>(self);
+  ssize_t index = fl_value_map_lookup_index(self, key);
```

Comment:
It might be worth mentioning in the header that there are no performance guarantees on map operations, since most people probably don't expect linear time. (I don't expect a lot of huge FlValue maps, but it wouldn't hurt to steer people gently away from trying.)

Reply:
I figured we could optimise this later if necessary, though I don't expect users to use FlValue for their core data structures :) I'll comment and add a FIXME.